### PR TITLE
Immutable collection with var are replaced by appropriate mutable

### PR DESCRIPTION
**Pull request by REASY**
- Immutable collection with var are replaced by appropriate mutable
- Added type for public methods and fields

### CODE REVIEW

File: src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceDriveIfAvailable.scala [hunk 1]
```diff
@@ -6,16 +6,18 @@ import beam.router.Modes.BeamMode.CAR
...
+
 /**
   * BEAM
   */
 class ModeChoiceDriveIfAvailable(val beamServices: BeamServices) extends ModeChoiceCalculator {
 
   def apply(alternatives: Seq[EmbodiedBeamTrip]): Option[EmbodiedBeamTrip] = {
-    var containsDriveAlt: Vector[Int] = Vector[Int]()
+    val containsDriveAlt: ArrayBuffer[Int] = ArrayBuffer[Int]()
     alternatives.zipWithIndex.foreach { alt =>
```
Comment (REASY) [inline src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceDriveIfAvailable.scala]: Fix this with:
```
alternatives.zipWithIndex.filter { case (beamTrip, idx) => beamTrip. tripClassifier == CAR }.map { case (beamTrip, idx) => idx}

```

**Comment by carloscaldas** [comment]
Since you are replacing immutable with mutable I recommend:
- Tight the visibility to private 
- Pay special attention to not share the references to other classes

**Comment by JustinPihony** [comment]
I agree with @carloscaldas As a rule of thumb you can use mutable internally, but expose immutable - otherwise it starts to get messy

File: src/main/java/beam/agentsim/agents/choice/logit/LatentClassChoiceModel.scala [hunk 2]
```diff
@@ -11,34 +11,35 @@ import org.supercsv.io.CsvBeanReader
...
 
     var row: LccmData = new LccmData()
-    var data: Vector[LccmData] = Vector()
+    val data = mutable.ArrayBuffer[LccmData]()
     while (beanReader.read[LccmData](row, header, processors: _*) != null) {
-      if (Option(row.value).isDefined && !row.value.isNaN) data = data :+ row.clone().asInstanceOf[LccmData]
+      if (Option(row.value).isDefined && !row.value.isNaN) data += row.clone().asInstanceOf[LccmData]
       row = new LccmData() // Turns out that if beanReader encounters a missing field, it just doesn't do the setField() rather than set it to null.... so we need to mannualy reset after every read
     }
     data
```
Comment (JustinPihony) [inline src/main/java/beam/agentsim/agents/choice/logit/LatentClassChoiceModel.scala]: data is still an ArrayBuffer under the covers, and technically mutable. I'd propose `data.toIndexedSeq`

File: src/main/scala/beam/agentsim/agents/Population.scala [hunk 3]
```diff
@@ -33,21 +33,20 @@ class Population(val scenario: Scenario, val beamServices: BeamServices, val sch
     case _: Exception => Stop
     case _: AssertionError => Stop
   }
-  private implicit val timeout = Timeout(50000, TimeUnit.SECONDS)
+  private implicit val timeout: Timeout = Timeout(50000, TimeUnit.SECONDS)
 
   import context.dispatcher
 
-  private var personToHouseholdId: Map[Id[Person], Id[Household]] = Map()
+  private val personToHouseholdId: mutable.Map[Id[Person], Id[Household]] = mutable.Map[Id[Person], Id[Household]]()
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/Population.scala]: As this is only done at creation then it seems the only reason mutability was used was because of the initialization. So, I'd argue for something more like this:

```
  private val personToHouseholdId: scala.collection.Map[Id[Person], Id[Household]] = 
    scenario.getHouseholds.getHouseholds.asScala
      .flatMap{ case (householdId, matSimHousehold) =>
        matSimHousehold.getMemberIds.asScala.map(_ -> householdId)
      }
```

Of course if you feel the performance of mutation to be better, then you could do something more like this and end with the same result:

```
  def initPersonToHouseholdId(houseHolds: java.util.Map[Id[Household], Household]): scala.collection.Map[Id[Person], Id[Household]] = {
    val personToHouseholdId: mutable.Map[Id[Person], Id[Household]] = mutable.Map[Id[Person], Id[Household]]()
    scenario.getHouseholds.getHouseholds.forEach { (householdId, matSimHousehold) =>
      personToHouseholdId ++= matSimHousehold.getMemberIds.asScala.map(personId => personId -> householdId)
    }
    personToHouseholdId.toMap
  }
  
  private val personToHouseholdId: scala.collection.Map[Id[Person], Id[Household]] = 
    initPersonToHouseholdId(scenario.getHouseholds.getHouseholds)
```

File: src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceDriveIfAvailable.scala (hunk 1 above)
Comment (JustinPihony) [reply src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceDriveIfAvailable.scala]: Even better would be:

`val containsDriveAlt = alternatives.zipWithIndex.collect{ case(trip, idx) if(trip.tripClassifier == CAR) => idx}`

File: src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceLCCM.scala [hunk 4]
```diff
@@ -136,7 +136,7 @@ class ModeChoiceLCCM(val beamServices: BeamServices, val lccm: LatentClassChoice
       ModeChoiceData(altAndIdx._1.tripClassifier, tourType, vehicleTime, walkTime, waitTime, bikeTime, totalCost.toDouble, altAndIdx._2)
     }
 
-    val groupedByMode = modeChoiceAlternatives.sortBy(_.mode.value).groupBy(_.mode)
+    val groupedByMode: Map[BeamMode, Seq[ModeChoiceData]] = modeChoiceAlternatives.groupBy(_.mode)
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceLCCM.scala]: I believe that the sort will indeed be lost in the long term, but was this meant to be an attempt to sort the values within the grouping? I don't see any problem with the optimization, and am just curious if the original intent was lost and should be fixed instead of optimized?

File: src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceRideHailIfAvailable.scala [hunk 5]
```diff
@@ -6,16 +6,18 @@ import beam.router.Modes.BeamMode.RIDE_HAIL
...
+import scala.collection.mutable.ArrayBuffer
+
 /**
   * BEAM
   */
 class ModeChoiceRideHailIfAvailable(val beamServices: BeamServices) extends ModeChoiceCalculator {
 
   override def apply(alternatives: Seq[EmbodiedBeamTrip]): Option[EmbodiedBeamTrip] = {
-    var containsRideHailAlt: Vector[Int] = Vector[Int]()
+    val containsRideHailAlt: ArrayBuffer[Int] = ArrayBuffer[Int]()
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceRideHailIfAvailable.scala]: Same as above:

`val containsRideHailAlt = alternatives.zipWithIndex.collect{ case(trip, idx) if(trip.tripClassifier == RIDE_HAIL) => idx}`

File: src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceTransitIfAvailable.scala [hunk 6]
```diff
@@ -5,18 +5,20 @@ import beam.router.Modes
...
 /**
   * BEAM
   */
 class ModeChoiceTransitIfAvailable(val beamServices: BeamServices) extends ModeChoiceCalculator {
 
   override def clone(): ModeChoiceCalculator = new ModeChoiceTransitIfAvailable(beamServices)
 
-  override def apply(alternatives: Seq[EmbodiedBeamTrip]) = {
-    var containsTransitAlt: Vector[Int] = Vector[Int]()
+  override def apply(alternatives: Seq[EmbodiedBeamTrip]): Option[EmbodiedBeamTrip] = {
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceTransitIfAvailable.scala]: Same as other alt comments. I'll stop here and this should just be repeated for any future instances like this.

File: src/main/scala/beam/agentsim/agents/household/HouseholdActor.scala [hunk 7]
```diff
@@ -140,32 +140,32 @@ object HouseholdActor {
...
       * Available [[Vehicle]]s in [[Household]].
       */
-    var _vehicles: Vector[Id[Vehicle]] = vehicles.keys.toVector.map(x => Id.createVehicleId(x))
+    val _vehicles: Vector[Id[Vehicle]] = vehicles.keys.toVector.map(x => Id.createVehicleId(x))
 
     /**
       * Concurrent [[MobilityStatusInquiry]]s that must receive responses before completing vehicle assignment.
       */
-    var _pendingInquiries: Map[Id[MobilityStatusInquiry], Id[Vehicle]] = Map[Id[MobilityStatusInquiry], Id[Vehicle]]()
+    val _pendingInquiries: Map[Id[MobilityStatusInquiry], Id[Vehicle]] = Map[Id[MobilityStatusInquiry], Id[Vehicle]]()
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/household/HouseholdActor.scala]: This one can be dropped. It isn't used.

File: src/main/scala/beam/agentsim/agents/rideHail/RideHailingManager.scala [hunk 8]
```diff
@@ -90,9 +90,9 @@ class RideHailingManager(
   }
 
   // TODO: remove following (replace by RideHailModifyPassengerScheduleManager)
-  val repositioningPassengerSchedule = mutable.Map[Id[Vehicle], (Id[Interrupt],Option[PassengerSchedule])]()
-  val reservationPassengerSchedule=mutable.Map[Id[Vehicle], (Id[Interrupt],ModifyPassengerSchedule)]()
-  val repositioningVehicles = mutable.Set[Id[Vehicle]]() // TODO: move to RideHailModifyPassengerScheduleManager?
+  val repositioningPassengerSchedule: mutable.Map[Id[Vehicle], (Id[Interrupt], Option[PassengerSchedule])] = mutable.Map[Id[Vehicle], (Id[Interrupt],Option[PassengerSchedule])]()
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/rideHail/RideHailingManager.scala]: Is there actually a bug here? It appears that the code which `put` values is commented out. So all of the contains that WERE here are going to be empty. Same for `reservationPassengerSchedule`

File: src/main/scala/beam/agentsim/agents/rideHail/RideHailingManager.scala [hunk 9]
```diff
@@ -859,14 +859,12 @@ class RideHailingManager(
...
-  def lockVehicle(vehicleId: Id[Vehicle]) = {
+  def lockVehicle(vehicleId: Id[Vehicle]): Unit = {
     lockedVehicles += vehicleId
   }
 
 
-  def getVehicleFuelLevel(vehicleId: Id[Vehicle]): Double = {
-    vehicleFuelLevel.get(vehicleId).get
-  }
+  def getVehicleFuelLevel(vehicleId: Id[Vehicle]): Double = vehicleFuelLevel(vehicleId)
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/rideHail/RideHailingManager.scala]: Should this return an Option? Granted this is not really used afaict. Just as an API example in `StanfordRideHailAllocationManagerV1`, so probably not a big deal...maybe even remove?

File: src/main/scala/beam/agentsim/agents/rideHail/allocationManagers/DefaultRideHailResourceAllocationManager.scala [hunk 10]
```diff
@@ -14,7 +14,7 @@ class DefaultRideHailResourceAllocationManager extends RideHailResourceAllocatio
   }
 
 // TODO RW/Asif: how to make sure no one ever can call this?
-def allocateVehicles(allocationsDuringReservation: Vector[(VehicleAllocationRequest, Option[VehicleAllocation])]): Vector[(VehicleAllocationRequest, Option[VehicleAllocation])] = {
+def allocateVehicles(allocationsDuringReservation: Vector[(VehicleAllocationRequest, Option[VehicleAllocation])]): IndexedSeq[(VehicleAllocationRequest, Option[VehicleAllocation])] = {
   log.error("batch processing is not implemented for DefaultRideHailResourceAllocationManager")
     //???
   return allocationsDuringReservation
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/rideHail/allocationManagers/DefaultRideHailResourceAllocationManager.scala]: What's the purpose of the signature change here? Underlying, the data is the same - same for other areas like this

File: src/main/scala/beam/agentsim/agents/rideHail/allocationManagers/StanfordRideHailAllocationManagerV1.scala [hunk 11]
```diff
@@ -59,7 +61,7 @@ class StanfordRideHailAllocationManagerV1(val rideHailingManager: RideHailingMan
         }
       }
 
-      result = result :+ (vehicleAllocationRequest, vehicleAllocation)
+      result += ((vehicleAllocationRequest, vehicleAllocation))
     }
     result
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/rideHail/allocationManagers/StanfordRideHailAllocationManagerV1.scala]: return `result.toIndexedSeq` to truly enforce the type

File: src/main/scala/beam/router/r5/R5RoutingWorker.scala [hunk 12]
```diff
@@ -280,8 +281,8 @@ class R5RoutingWorker(val beamServices: BeamServices, val transportNetwork: Tran
           //TODO make a more sensible window not just 30 minutes
           startTime >= time.fromTime && startTime <= time.fromTime + 1800
         }.map(itinerary => {
-          var legsWithFares = Vector[(BeamLeg, Double)]()
-          maybeWalkToVehicle.foreach(legsWithFares +:= (_, 0.0))
+          val legsWithFares = ArrayBuffer[(BeamLeg, Double)]()
```
Comment (JustinPihony) [inline src/main/scala/beam/router/r5/R5RoutingWorker.scala]: How about just move it into a one line initializer?

`val legsWithFares = maybeWalkToVehicle.map(x=> ArrayBuffer((x, 0.0))).getOrElse(ArrayBuffer[(BeamLeg, Double)]())`

**Comment by REASY** [comment]
@carloscaldas, @JustinPihony , thanks for the raised questions and I think all of them should be solved. But I'm not sure that it should be in this PR. This PR is about changing immutable types to mutable. I did not want to create extra complexity to figure out what I broke by doing a lot of changes in on PR, if I would break anything.

**Comment by colinsheppard** [comment]
I concur with @JustinPihony in that we should treat immutability as a means to an end, not an end in itself. 

Internal to a method, there's no harm in using mutability for convenience (particularly convenient for devs who are still in the habit of procedural programming).  Not to say that immutability is worse, it's just not necessarily worth a lot of time refactoring or fussing over.

Internal to an Actor, immutability is also acceptable in my opinion. Everyone knows (or should know) never to attempt to directly access an Actor's internal state, so the Actor should feel free to mutate state in whatever way is most concise, understandable, and maintainable. All that said, I have been convinced by @michaz that using the Akka FMS state data as the way to manage state does have benefit because that state is passed around, put into logging, etc, which greatly facilitates debugging. It also has the benefit of making it easier to mock Actors for testing.

So, in summary, I'm happy to see this work going on, but I think the key emphasis should be put into places where we were managing global state but using mutability for convenience (BeamServices in particular), a regrettable choice made early on (by me) when I was new to Scala. We should correct this and find a way to make BeamServices immutable or make all mutable fields private and then disallow changes other than during initialization.

**Comment by carloscaldas** [comment]
Immutability can be expensive: even if we we don't have to recreate the
whole object still have have to allocate memory again (even for a small
object is expensive)

I would sumarize in the following way:

-object does not leave method scope and don't make parallel
computation: *mutable
object*
-object leaves the class scope: *immutable object*
-object leaves the method scope: *immutability sometimes (should analyse
case by case)*
*-makes parallel computation over the object: immutable *(this is not a
silver bullet but will make better code in 90% of the times)

File: src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceLCCM.scala (hunk 4 above)
Comment (REASY) [reply src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceLCCM.scala]: It's hard to say about original intent, maybe @colinsheppard can help here.

File: src/main/scala/beam/agentsim/agents/rideHail/allocationManagers/DefaultRideHailResourceAllocationManager.scala (hunk 10 above)
Comment (REASY) [reply src/main/scala/beam/agentsim/agents/rideHail/allocationManagers/DefaultRideHailResourceAllocationManager.scala]: Using `IndexedSeq` allows to fill `ArrayBuffer` and return it as result, instead of converting it to vector by calling `toVector`

File: src/main/java/beam/agentsim/agents/choice/logit/LatentClassChoiceModel.scala (hunk 2 above)
Comment (REASY) [reply src/main/java/beam/agentsim/agents/choice/logit/LatentClassChoiceModel.scala]: If I get this right [TraversableOnce.scala](https://github.com/scala/scala/blob/6ac6da8b61a3a427089a166c7802a940eac71064/src/library/scala/collection/TraversableOnce.scala#L308), `toIndexedSeq` will allocate new buffer and copy elements there - more GC pressure.

File: src/main/scala/beam/agentsim/agents/rideHail/allocationManagers/StanfordRideHailAllocationManagerV1.scala (hunk 11 above)
Comment (REASY) [reply src/main/scala/beam/agentsim/agents/rideHail/allocationManagers/StanfordRideHailAllocationManagerV1.scala]: If I get this right [TraversableOnce.scala](https://github.com/scala/scala/blob/6ac6da8b61a3a427089a166c7802a940eac71064/src/library/scala/collection/TraversableOnce.scala#L308), toIndexedSeq will allocate new buffer and copy elements there - more GC pressure.
